go/worker/compute/executor/committee/node.go

```diff
@@ -417,13 +496,239 @@ func (n *Node) HandleNewBlockLocked(blk *block.Block) {
 			"round", blk.Header.Round,
 			"header_hash", blk.Header.EncodedHash(),
 		)
+		// Removed processed transactions from queue.
+		if err := n.removeTxBatch(state.raw); err != nil {
+			n.logger.Warn("failed removing processed batch from queue",
+				"err", err,
+				"batch", state.raw,
+			)
+		}
+
 		n.transitionLocked(StateWaitingForBatch{})
 
 		// Record time taken for successfully processing a batch.
 		batchProcessingTime.With(n.getMetricLabels()).Observe(time.Since(state.batchStartTime).Seconds())
 	}
 }
 
+// checkTx checks the given call in the node's runtime.
+func (n *Node) checkTx(ctx context.Context, call []byte) error {
+	n.commonNode.CrossNode.Lock()
+	currentBlock := n.commonNode.CurrentBlock
+	n.commonNode.CrossNode.Unlock()
+
+	if currentBlock == nil {
+		return errNotReady
+	}
+
+	checkRq := &protocol.Body{
+		RuntimeCheckTxBatchRequest: &protocol.RuntimeCheckTxBatchRequest{
+			Inputs: transaction.RawBatch{call},
+			Block:  *currentBlock,
+		},
+	}
+	rt := n.GetHostedRuntime()
+	if rt == nil {
+		n.logger.Error("CheckTx: hosted runtime not initialized")
+		return errNotReady
+	}
+	resp, err := rt.Call(ctx, checkRq)
+	switch {
+	case err != nil:
+		n.logger.Error("CheckTx: runtime call error",
+			"err", err,
+		)
+		return err
+	case resp.RuntimeCheckTxBatchResponse == nil:
+		n.logger.Error("CheckTx: runtime response is nil")
+		return errCheckTxFailed
+	case resp.RuntimeCheckTxBatchResponse.Results == nil:
+		n.logger.Error("CheckTx: response contains no results")
+		return errCheckTxFailed
+	case len(resp.RuntimeCheckTxBatchResponse.Results) != 1:
+		n.logger.Error("CheckTx: runtime response doesn't contain exactly one result",
+			"num_results", len(resp.RuntimeCheckTxBatchResponse.Results),
+		)
+		return errCheckTxFailed
+	}
+
+	// Interpret CheckTx result.
+	resultRaw := resp.RuntimeCheckTxBatchResponse.Results[0]
+	var result transaction.TxnOutput
+	if err = cbor.Unmarshal(resultRaw, &result); err != nil {
+		n.logger.Error("CheckTx: runtime response failed to deserialize",
+			"err", err,
+		)
+		return errCheckTxFailed
+	}
+	if result.Error != nil {
+		n.logger.Error("CheckTx: runtime failed with error",
+			"err", result.Error,
+		)
+		return fmt.Errorf("%w: %s", errCheckTxFailed, *result.Error)
+	}
+
+	return nil
+}
+
+// QueueTx queues a runtime transaction for scheduling.
+func (n *Node) QueueTx(call []byte) error {
+	n.schedulerMutex.RLock()
+	defer n.schedulerMutex.RUnlock()
+
+	if n.scheduler == nil || !n.scheduler.IsInitialized() {
+		return errNotReady
+	}
+	if err := n.scheduler.ScheduleTx(call); err != nil {
+		return err
+	}
+
+	incomingQueueSize.With(n.getMetricLabels()).Set(float64(n.scheduler.UnscheduledSize()))
+
+	return nil
+}
+
+// removeTxBatch removes a batch from scheduling queue.
+func (n *Node) removeTxBatch(batch [][]byte) error {
+	// XXX: remove batch can only happen after a batch was already scheduled, meaning
```

Comment:
This is fine as long as there is no way for the scheduler to change (e.g., due to a runtime descriptor update) -- currently there is no way for it to change once set. In that case we would want to hold the mutex here.

Reply:
Yeah that's what we also have in the comment in `schedulerMutex` decleration, and since we explicitly state that there i decided to follow that. However i'm also fine with leaving that here for future proofness.